components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -336,189 +604,254 @@ impl RangeManager {
+    pub(crate) fn record_in_region_being_written(
         &mut self,
         write_batch_id: u64,
-        range: &CacheRange,
+        range: CacheRange,
     ) {
-        self.ranges_being_written
+        self.regions_being_written
             .entry(write_batch_id)
             .or_default()
-            .push(range.clone())
+            .push(range);
     }
 
-    pub(crate) fn clear_ranges_in_being_written(
+    pub(crate) fn clear_regions_in_being_written(
         &mut self,
         write_batch_id: u64,
         has_entry_applied: bool,
     ) {
-        let ranges = self.ranges_being_written.remove(&write_batch_id);
+        let regions = self.regions_being_written.remove(&write_batch_id);
         if has_entry_applied {
-            assert!(!ranges.unwrap().is_empty());
+            assert!(!regions.unwrap().is_empty());
         }
     }
 
-    pub fn on_gc_finished(&mut self, range: BTreeSet<CacheRange>) {
-        assert_eq!(range, std::mem::take(&mut self.ranges_in_gc));
+    pub fn load_region(&mut self, region: Region) -> Result<(), LoadFailedReason> {
+        use RegionState::*;
+        if let Some(state) = self.check_overlap_with_region(&region) {
+            let reason = match state {
+                Pending | ReadyToLoad | Loading => LoadFailedReason::PendingRange,
+                Active => LoadFailedReason::Overlapped,
+                LoadingCanceled | PendingEvict | Evicting => LoadFailedReason::Evicting,
+            };
+            return Err(reason);
+        }
+        let meta = RangeMeta::new(region);
+        self.new_region_meta(meta);
+        Ok(())
     }
 
-    pub fn load_range(&mut self, cache_range: CacheRange) -> Result<(), LoadFailedReason> {
-        if self.overlap_with_range(&cache_range) {
-            return Err(LoadFailedReason::Overlapped);
-        };
-        if self.overlap_with_pending_range(&cache_range) {
-            return Err(LoadFailedReason::PendingRange);
+    // return `true` is the region is evicted.
+    pub(crate) fn split_region(
+        &mut self,
+        source_region: &Region,
+        mut new_regions: Vec<kvproto::metapb::Region>,
+    ) {
+        if let Some(region_meta) = self.region_meta(source_region.id) {
+            // if region is evicting, skip handling split for simplicity.
+            if region_meta.state >= RegionState::LoadingCanceled {
+                info!("region is evicted, skip split"; "meta" => ?&region_meta, "new_regions" => ?new_regions);
+                return;
+            }
+        } else {
+            info!("split source region not cached"; "region_id" => source_region.id);
+            return;
         }
-        if self.overlap_with_range_in_gc(&cache_range) {
-            return Err(LoadFailedReason::InGc);
+
+        let region_meta = self.remove_region(source_region.id);
+        assert!(region_meta.state <= RegionState::Active);
+        if region_meta.region.get_region_epoch().version != source_region.get_region_epoch().version
```

Comment:
Source here means region before split or after? In which cases the version equals and not equals?

Reply:
`Source` is region before split. So if the region is in "ReadyToLoad, Loading, Active" state, the source region epoch should equal to the cached region epoch.